Enable multiple mime support with edit & restore flows

mehulkaklotar opened the pull request:
This PR adds support for multiple mime types when image is edited and restored.

mehulkaklotar committed:
initial changes for save image for multiple mime type support

mehulkaklotar committed:
_wp_attachment_backup_sources meta store when image edited and saved

mehulkaklotar committed:
restore full size image source data when restored image

mehulkaklotar committed:
removed unwanted file changes commited

mehulkaklotar committed:
edit image - meta update sources and sizes and generate mime types

mehulkaklotar committed:
edit image - meta update backup sources fix

mehulkaklotar committed:
test cases - test_backup_the_sources_structure_alongside_the_full_size added

jjgrainger commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1038,177 @@ function wp_save_image( $post_id ) {
+					&& isset( $meta['file'] )
+					&& 0 !== strpos( $meta['file'], '/' )
+					&& ':\\' !== substr( $meta['file'], 1, 2 )
+			) {
```
Can we reduce the indentation here?

```suggestion
			if (
				null === $image_directory
				&& isset( $meta['file'] )
				&& 0 !== strpos( $meta['file'], '/' )
				&& ':\\' !== substr( $meta['file'], 1, 2 )
			) {
```

jjgrainger commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1038,177 @@ function wp_save_image( $post_id ) {
+				$meta['sizes'][ $size_name ]['sources'][ $targeted_mime ] = array(
+						'file'     => $subsized_images[ $targeted_mime ][ $size_name ]['file'],
+						'filesize' => wp_filesize( $subsize_path ),
+				);
```
Same here, can we reduce the indentation?

```suggestion
				$meta['sizes'][ $size_name ]['sources'][ $targeted_mime ] = array(
					'file'     => $subsized_images[ $targeted_mime ][ $size_name ]['file'],
					'filesize' => wp_filesize( $subsize_path ),
				);
```

jjgrainger commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1038,177 @@ function wp_save_image( $post_id ) {
+				$current_extension = pathinfo( $thumbnail_file, PATHINFO_EXTENSION );
+
+				// Create a file with then new extension out of the targeted file.
+				$target_file_name     = preg_replace( "/\.$current_extension$/", ".$extension", $thumbnail_file );
```
Would `str_replace` be suitable here, I believe it's faster?

mehulkaklotar replied:
@jjgrainger It is because we wanted find the last part of path which is extension and not the any part in between. Plugin has this changes already in place. https://github.com/WordPress/performance/blob/trunk/modules/images/webp-uploads/image-edit.php#L188

jjgrainger commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1038,177 @@ function wp_save_image( $post_id ) {
+			continue;
+		}
+
+		if (
+				isset( $meta['sizes'][ $size_name ] ) && ! empty( $meta['sizes'][ $size_name ] ) &&
+				$meta['sizes'][ $size_name ]['file'] !== $old_metadata['sizes'][ $size_name ]['file']
+		) {
```
Can we reduce the indentation here?

```suggestion
		if (
			isset( $meta['sizes'][ $size_name ] ) && ! empty( $meta['sizes'][ $size_name ] ) &&
			$meta['sizes'][ $size_name ]['file'] !== $old_metadata['sizes'][ $size_name ]['file']
		) {
```

jjgrainger reviewed (changes requested):
Thanks @mehulkaklotar 

Left a few minor comments, but overall looks good to me

mehulkaklotar committed:
Update src/wp-admin/includes/image-edit.php

Co-authored-by: Joe Grainger <904708+jjgrainger@users.noreply.github.com>

mehulkaklotar committed:
Update src/wp-admin/includes/image-edit.php

Co-authored-by: Joe Grainger <904708+jjgrainger@users.noreply.github.com>

mehulkaklotar committed:
Update src/wp-admin/includes/image-edit.php

Co-authored-by: Joe Grainger <904708+jjgrainger@users.noreply.github.com>

mehulkaklotar committed:
fix restore image backup sources not saving

mehulkaklotar committed:
variable name changes in backup sources next key target

mehulkaklotar committed:
test case - test_restore_the_sources_array_from_the_backup_when_an_image_is_edited added

mehulkaklotar committed:
fix phpcs errors

audrasjb committed:
Themes: Add a hook to filter theme header image URL.

This changeset introduces the `get_header_image` filter, which can be used to modify header image URL returned by `get_header_image()`, in themes that support the Header Image feature. 

Props hztyfoon, audrasjb, mukesh27, SergeyBiryukov, costdev.
Fixes #56180.


git-svn-id: https://develop.svn.wordpress.org/trunk@53741 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Add missing function description in `wp-admin/includes/comment.php`.

See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53742 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Use third-person singular verbs for function descriptions in `wp-admin/includes/comment.php`, as per docblocks standards.

See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53743 602fd350-edb4-49c9-b593-d223f7449a82

peterwilsoncc committed:
Embeds: Add Pocket Casts as a trusted oEmbed provider.

Props mattwondra, peterwilsoncc, pento, audrasjb.
Fixes #55860.



git-svn-id: https://develop.svn.wordpress.org/trunk@53744 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Administration: Add labels to read-only form fields.

This changeset improves admin forms accessibility by adding labels to the following read-only form fields:

- Network setup screen: new visible label to the four textareas for code users need to paste into their wp-config file and the server configuration file (web.config or .htaccess).
- `setup-config.php`: new visible label to one textarea for code to include in the `wp-config` file manually.
- Admin toolbar: adds an `arial-label` attribute to the old "shortlink" feature (not used anymore but still activable by plugins).

Props sabernhardt, audrasjb, ryokuhi, joedolson.
Fixes #54302.


git-svn-id: https://develop.svn.wordpress.org/trunk@53745 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Add missing function description in `wp-admin/includes/template.php`.

This changeset also includes a few docblock standards fixes in the same file.

See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53746 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Coding Standards: Adjust the logic slightly in `wp_ajax_replyto_comment()`.

This aims to bring more consistency to the function logic by separating the comment content and type checks from the user existence check.

Follow-up to [8720], [9098], [29758].

See #55647.

git-svn-id: https://develop.svn.wordpress.org/trunk@53747 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Update `do_action()` docblock code example.

This changeset fixes an issue with the code example provided: `do_action()` doesn't return any value.

Props leogermani, thakkarhardik.
Fixes #55977.


git-svn-id: https://develop.svn.wordpress.org/trunk@53748 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Coding Standards: Move `wp-includes/wp-db.php` to `wp-includes/class-wpdb.php`.

This renames the file containing the `wpdb` class to conform to the coding standards.

This commit also includes:
* A new `wp-db.php` that loads the new file, for anyone that may have been including the file directly.
* Replacing references to the old filename with the new filename.

Fixes #56268. See #55647.

git-svn-id: https://develop.svn.wordpress.org/trunk@53749 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Coding Standards: Replace the old `wp-db.php` filename in `phpcompat.xml.dist`.

This ensures the correct file is allowlisted for use of `mysql_` extension in PHP < 7.0.

Follow-up to [53749].

See #56268, #55647.

git-svn-id: https://develop.svn.wordpress.org/trunk@53750 602fd350-edb4-49c9-b593-d223f7449a82

adamsilverstein committed:
Media: enable generating multiple mime types for image uploads; specifically WebP versions for JPEG images by default.

This changeset adds the capability for core media uploads to generate sub sized images in more than a single mime type. The output formats for each mime type can be controlled through a filter. WebP is used as an additional output format for JPEG images by default to improve front end performance.

When generating additional mime types, only images which are smaller than the respective original are retained. By default, additional mime type images are only generated for the built-in core image sizes and any custom sizes that have opted in.

Image meta is updated with a new 'sources' array containing file details for each mime type. Each image size in the 'sizes' array also gets a new 'sources' array that contains the image file details for each mime type.

This change also increases image upload retries to accommodate additional image sizes. It also adds a `$mime_type` parameter to the `wp_get_missing_image_subsizes` function and filter.

This change adds three new filters to enable full control of secondary mime image generation and output:

* A new filter `wp_image_sizes_with_additional_mime_type_support` that filters the sizes that support secondary mime type output. Developers can use this to control the output of additional mime type sub-sized images on a per size basis.
* A new filter `wp_upload_image_mime_transforms` that filters the output mime types for a given input mime type. Developers can use this to control generation of additional mime types for a given input mime type or even override the original mime type.
* A new filter `wp_content_image_mimes` which controls image mime type output selection and order for frontend content. Developers can use this to control the mime type output preference order for content images. Content images inserted from the media library will use the available image versions based on the order from this filter.

Thanks to the many contributors who helped develop, test and give feedback on this feature.

A haiku to summarize:

Upload a JPEG
Images of all sizes
Output as WebPs

Props flixos90, MatthiasReinholz, studiolxv, markhowellsmead, eatingrules, pbiron, mukesh27, joegrainger, mehulkaklotar, tweetythierry, akshitsethi, peterwilsoncc, eugenemanuilov, mitogh, shetheliving, clarkeemily, codekraft, mikeschroder, clorith, kasparsd, spacedmonkey, trevorpfromsandee, jb510, scofennellgmailcom, seedsca, cagsmith, karinclimber, dainemawer, baxbridge, grapplerulrich, sobatkras, chynnabenton, tonylocalword, barneydavey, kwillmorth, garymatthews919, olliejones, imarkinteractive, jeffpaul, feastdesignco, webbeetle, masteradhoc.

See #55443.



git-svn-id: https://develop.svn.wordpress.org/trunk@53751 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Application Passwords: Link a more accurate documentation page in User profile edit screen.

This changeset replaces a HelpHub link with more accurate DevHub link about environment types in User profile edit screen.

Props greg24, whaze.
Fixes #56267.


git-svn-id: https://develop.svn.wordpress.org/trunk@53752 602fd350-edb4-49c9-b593-d223f7449a82

adamsilverstein committed:
Upgrade/Install: track php extensions and image library support for WebP and AVIF.

Add the loaded php extensions as well as whether the server's image libraries support WebP and AVIF image formats to the data payload sent during upgrade checks. Collecting this data with the WordPress.org API will help the project make more data driven decisions about feature support. Note that all data can still be filtered with the `core_version_check_query_args` filter for privacy reasons.

Props dd32, SergeyBiryukov, mikeschroder, pbiron. 
Fixes #48116.



git-svn-id: https://develop.svn.wordpress.org/trunk@53753 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Formatting: Normalize to Unicode NFC encoding before converting accent characters in `remove_accents()`.

This changeset adds Unicode sequence normalization from NFD to NFC, via the `normalizer_normalize()` PHP function which is available with the recommended `intl` PHP extension.

This fixes an issue where NFD characters were not properly sanitized. It also provides a unit test for NFD sequences (alternate Unicode representations of the same characters).

Props NumidWasNotAvailable, targz, nacin, nunomorgadinho, p_enrique, gitlost, SergeyBiryukov, markoheijnen, mikeschroder, ocean90, pento, helen, rodrigosevero, zodiac1978, ironprogrammer, audrasjb, azaozz, laboiteare, nuryko, virgar, dxd5001, onnimonni, johnbillion.
Fixes #24661, #47763, #35951.
See #30130, #52654.


git-svn-id: https://develop.svn.wordpress.org/trunk@53754 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Coding Standards: Check if the `_deprecated_file()` function exists in `wp-db.php`.

This avoids a fatal error if the file is included directly outside of WordPress core, e.g. by HyperDB.

Additionally, replace `WPINC` with `wp-includes`, as the constant may not be defined yet in that context.

Follow-up to [53749], [53750].

See #56268, #55647.

git-svn-id: https://develop.svn.wordpress.org/trunk@53755 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Coding Standards: Use `__DIR__` instead of `ABSPATH` in `wp-db.php`.

If the file is included directly outside of WordPress core, `ABSPATH` may not be defined.

Follow-up to [53749], [53750], [53755].

See #56268, #55647.

git-svn-id: https://develop.svn.wordpress.org/trunk@53756 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Provide a more accurate description for `wp_is_theme_directory_ignored()`.

This function doesn't filter theme directories, it rather determines whether a directory should be ignored during export.

Props audrasjb, nareshbheda.
Fixes #56257.
See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53757 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Administration: Correct the escaping in documentation lookup for plugin and theme editor.

This brings consistency to similar code fragments that evolved in slightly different ways over the years:

* A check that the list of functions is not empty now exists in both editors.
* "Function Name..." is an option label, not an attribute, so `esc_html__()` is the correct function here.
* `esc_attr( urlencode( $function ) )` in the theme editor is replaced with `esc_attr( $function )`.
* `htmlspecialchars( $function )` in the theme editor is replaced with `esc_html( $function )`.

Follow-up to [10607], [10879], [11110], [11173], [11204], [11671], [14989].

Props aristath, justinahinon, poena, audrasjb, pratiweb, SergeyBiryukov.
Fixes #56007.

git-svn-id: https://develop.svn.wordpress.org/trunk@53758 602fd350-edb4-49c9-b593-d223f7449a82

spacedmonkey committed:
REST API: Use `wp_get_lastest_revision_id_and_total_count` function in `WP_REST_Posts_Controller` class. 

Add new function called `wp_get_lastest_revision_id_and_total_count`, that performs an optimized query to get the last revision and total and use it in `WP_REST_Posts_Controller` class. 

Props Spacedmonkey, timothyblynjacobs, furi3r, peterwilsoncc.
Fixes #55857.

git-svn-id: https://develop.svn.wordpress.org/trunk@53759 602fd350-edb4-49c9-b593-d223f7449a82

spacedmonkey committed:
REST API: Avoid unnecessarily preparing item links.

Do not call the `prepare_links` methods in core REST API controllers, unless the `_links` or `_embedded` fields are requested. There is no need to prepare links if they are never returned in the response. This saves resources, as many calls to `prepare_links` methods perform database queries. 

Props Spacedmonkey, timothyblynjacobs, rachelbaker, desrosj, dlh, hellofromTonya.
Fixes #52992.

git-svn-id: https://develop.svn.wordpress.org/trunk@53760 602fd350-edb4-49c9-b593-d223f7449a82

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -799,6 +801,32 @@ function wp_restore_image( $post_id ) {
 		}
 	}
 
+	$next_full_size_key_from_backup = null;
```
it would be nice to document what this variable is for

desrosj committed:
Twenty Twenty-One: Update NPM dependencies.

This updates the NPM dependencies for the Twenty Twenty-One theme to the latest versions.

Modifications to the built files are included in this commit, which are a result of updating the `sass` dependency to the latest version.

Previously, trailing loud comments (`/* ... */`) were pushed to the next line,. Now the comment location is preserved, resulting in the built CSS files more closely resembling the theme’s SCSS files.

See #55652.

git-svn-id: https://develop.svn.wordpress.org/trunk@53761 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Bundled Themes: Update NPM dependencies for Twenty Twenty and Twenty Nineteen.

There are no changes to any built files after updating.

See #55652.

git-svn-id: https://develop.svn.wordpress.org/trunk@53762 602fd350-edb4-49c9-b593-d223f7449a82

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+	$resize_sizes = array();
+	foreach ( $old_metadata['sizes'] as $size_name => $size_details ) {
+		// If the target is 'nothumb', skip generating the 'thumbnail' size.
+		if ( 'nothumb' === $target && 'thumbnail' === $size_name ) {
```
where does "$target" come from?

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+	$main_images        = array();
+	$subsized_images    = array();
+
+	$transforms = wp_upload_image_mime_transforms( $post_id, false );
```
wp_upload_image_mime_transforms only takes a single parameter

mehulkaklotar replied:
Oh! nice catch. It was in previous version of add/webp-uploads branch, now it is removed.

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+				continue;
+			}
+
+			$extension = explode( '|', $allowed_mimes[ $targeted_mime ] );
```
could you use wp_get_default_extension_for_mime_type?

mehulkaklotar replied:
good suggestion. I have now used the function to get default mime type here.

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+
+				$subsized_images[ $targeted_mime ] = array( 'thumbnail' => $result );
+			} else {
+				$destination = trailingslashit( $original_directory ) . "{$filename}.{$extension}";
```
i wonder if you could be using $img->generate_filename() here?

mehulkaklotar replied:
here the destination file name is already there from the original file name so we just creating path where to save the mimed version of the image.

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+			$image_directory = null;
+			if ( isset( $main_images[ $targeted_mime ]['path'], $main_images[ $targeted_mime ]['file'] ) && file_exists( $main_images[ $targeted_mime ]['path'] ) ) {
+				// Add sources to original image metadata.
+				$meta['sources'][ $targeted_mime ] = array(
```
I think you could use `_wp_get_sources_from_meta` here

mehulkaklotar replied:
we can't use the `_wp_get_sources_from_meta` here as we are not using meta to get sources from, instead we are generating new sources in the meta from the main images.

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+				}
+			}
+
+			if ( null === $image_directory ) {
```
why all the work to find image_directory? I think you can just let the img class decide as the section above?

mehulkaklotar replied:
`$image_directory` is to get where that particular image size is located and we can save it in image source meta. so we will need to get it from the image file data.

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+				$subsized_images[ $targeted_mime ] = array( 'thumbnail' => $result );
+			} else {
+				$destination = trailingslashit( $original_directory ) . "{$filename}.{$extension}";
+				$result      = $img->save( $destination, $targeted_mime );
```
is this for the full sized images? I think you can just do |
$img->set_output_mime_type ( $targeted_mime );
$img->save();

save should then construct the correct filename and you won't need any of the code above

mehulkaklotar replied:
I have committed changes for this in latest commit.

adamsilverstein commented on tests/phpunit/tests/image/functions.php:
```diff
@@ -248,6 +248,115 @@ public function test_is_gd_image_invalid_types() {
+	/**
+	 * Backup the sources structure alongside the full size
+	 *
+	 * @test
```
reference the ticket - `@ticket 55443`

adamsilverstein commented on tests/phpunit/tests/image/functions.php:
```diff
@@ -248,6 +248,115 @@ public function test_is_gd_image_invalid_types() {
+		require_once ABSPATH . 'wp-admin/includes/image-edit.php';
+
+		$filename = DIR_TESTDATA . '/images/canola.jpg';
+		$contents = file_get_contents( $filename );
```
you should be able to use `_make_attachment` to simplify this (or maybe self::factory()->attachment->create_object)

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+				// Create a file with then new extension out of the targeted file.
+				$target_file_name     = preg_replace( "/\.$current_extension$/", ".$extension", $thumbnail_file );
+				$target_file_location = path_join( $original_directory, $target_file_name );
+				$result               = $img->save( $target_file_location, $targeted_mime );
```
can you add inline docs indicating which file is being created in each save spot? 

I don't think you need to construct the filename, you can use `set_output_mime_type` and the image editor class will generate the correct filename.

mehulkaklotar replied:
this particular block of code, generates/saves thumbnail images as it is in the condition that checks if `'thumbnail' === $target`. In this we can't use just `set_output_mime_type` as it will use image file of full size instead of thumbnail and will generate that file only.

adamsilverstein commented on src/wp-admin/includes/image-edit.php:
```diff
@@ -1026,12 +1061,177 @@ function wp_save_image( $post_id ) {
+				// Create a file with then new extension out of the targeted file.
+				$target_file_name     = preg_replace( "/\.$current_extension$/", ".$extension", $thumbnail_file );
+				$target_file_location = path_join( $original_directory, $target_file_name );
+				$result               = $img->save( $target_file_location, $targeted_mime );
```
Before saving, so we need to consider deleting files? the block above does and checks the IMAGE_EDIT_OVERWRITE constant

mehulkaklotar replied:
Fixed deleting attachment sources files when constant is true.

spacedmonkey committed:
Cache API: Add `wp_cache_flush_group` function. 

Add a new plugable function called `wp_cache_flush_group`, that will allow developers to clear whole cache groups with a single call. Developers can detect if their current implementation of an object cache supports flushing by group, by calling `wp_cache_supports_group_flush` which returns true if it is supported. If the developers of the object cache drop-in has not implemented `wp_cache_flush_group` and `wp_cache_supports_group_flush`, these functions are polyfilled and `wp_cache_supports_group_flush` defaults to false.

Props Spacedmonkey, filosofo, ryan, sc0ttkclark, SergeyBiryukov, scribu, Ste_95, dd32, dhilditch, dougal, lucasbustamante, dg12345, tillkruess, peterwilsoncc, flixos90, pbearne.
Fixes #4476.

git-svn-id: https://develop.svn.wordpress.org/trunk@53763 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Media: Allow filtering audio file metadata in `wp_read_audio_metadata()`.

This changeset introduces the `wp_read_audio_metadata` hook which allows to filter metadata extracted from an uploaded audio file.

This brings consistency with corresponding filters available for image and video file types:

- `wp_read_image_metadata` added in [6313] / #5162.
- `wp_read_video_metadata` added in [41746] / #35218.

Props luigipulcini, SergeyBiryukov, mukesh27.
Fixes #55828.


git-svn-id: https://develop.svn.wordpress.org/trunk@53764 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Fix indentation issues in `wp_read_video_metadata` and `wp_read_audio_metadata` docblocks.

Follow-up to [53764].

See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53765 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Add function description and `@since` mention for `get_upload_iframe_src()`.

Follow-up to [14015].

See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53766 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Cache API: Make the placement of `wp_cache_flush_group()` more consistent.

Includes:
* Placing `WP_Object_Cache::flush_group()` next to `::flush()`.
* Placing `wp_cache_supports_group_flush()` next to `wp_cache_flush_group()`.
* Placing the `wp_cache_flush_group()` unit test next to the `::flush()` method test.
* Removing test name from assertion messages, as it is already mentioned directly above in case of failure.
* Adjusting function descriptions per the documentation standards.

Follow-up to [52706], [53763].

See #55647, #4476.

git-svn-id: https://develop.svn.wordpress.org/trunk@53767 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Docs: Document that the `$file_format` parameter of metadata filters can be null.

This affects the `wp_read_video_metadata` and `wp_read_audio_metadata` filters.

Follow-up to [41746], [53764].

See #55646, #55828.

git-svn-id: https://develop.svn.wordpress.org/trunk@53768 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Revisions: Correct the function name for retrieving the last revision ID and total count.

Includes:
* Renaming the function to `wp_get_last_revision_id_and_total_count()`.
* Making the default value for `$post` consistent with `wp_get_post_revisions()`.
* Making `WP_Error` codes more specific and using them in test assertions.
* Adjusting the function description per the documentation standards.

Follow-up to [53759].

Props JustinSainton, SergeyBiryukov.
See #55857.

git-svn-id: https://develop.svn.wordpress.org/trunk@53769 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Docs: List the expected type first in a few functions:

* `is_allowed_http_origin()`
* `doing_filter()`
* `wp_get_post_revisions_url()`

Follow-up to [28010], [28889], [30674], [46696], [47060], [48068], [49929], [49963], [52095], [51286], [52111].

See #55646.

git-svn-id: https://develop.svn.wordpress.org/trunk@53770 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Posts, Post Types: Check if the post type exists in `wp_insert_post()`.

This avoids an `Attempt to read property "cap" on null` PHP warning when checking an unregistered post type's `publish_posts` capability to disallow contributors setting the post slug for pending posts.

Follow-up to [9055], [42380].

Props Chouby, mukesh27, rafiahmedd, SergeyBiryukov.
Fixes #55877.

git-svn-id: https://develop.svn.wordpress.org/trunk@53771 602fd350-edb4-49c9-b593-d223f7449a82

peterwilsoncc committed:
REST API: Prime post caches in comments endpoint.

Prevent each call to `get_post()` triggering a seperate database query by priming the cache of each comment's post in 
the REST API's comments endpoint.

Props spacedmonkey, timothyblynjacobs.
Fixes #56272.



git-svn-id: https://develop.svn.wordpress.org/trunk@53772 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Tests: Add failure messages for site icon and custom logo tests.

This makes the assertions more helpful, as per the [https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions Writing PHP Tests] guidelines:
> All PHPUnit assertions, as well as all WordPress custom assertions, allow for a `$message` parameter to be passed. This message will be displayed when the assertion fails and can help immensely when debugging a test. This parameter should always be used if more than one assertion is used in a test method.

Follow-up to [33181], [36905].

See #55652.

git-svn-id: https://develop.svn.wordpress.org/trunk@53773 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Tests: Declare `custom-logo` theme support for custom logo tests.

This addresses failures in `has_custom_logo()` and `get_custom_logo()` tests when being run separately:
{{{
1) Tests_General_Template::test_has_custom_logo
Custom logo should not be set after removal.
Failed asserting that true is false.
tests/phpunit/tests/general/template.php:291

2) Tests_General_Template::test_get_custom_logo
Custom logo should not be set after removal.
Failed asserting that a string is empty.
tests/phpunit/tests/general/template.php:336
}}}

Specifically, this ensures that the `site_logo` option gets deleted in `_delete_site_logo_on_remove_theme_mods()`, which in turn prevents the `core/site-logo` block filters from affecting the custom logo tests.

Alternatively, these filters could be removed instead:
{{{
remove_filter( 'theme_mod_custom_logo', '_override_custom_logo_theme_mod' );
remove_filter( 'pre_set_theme_mod_custom_logo', '_sync_custom_logo_to_site_logo' );
}}}

Follow-up to [36905], [51091], [51421], [52042].

See #55652.

git-svn-id: https://develop.svn.wordpress.org/trunk@53774 602fd350-edb4-49c9-b593-d223f7449a82

mehulkaklotar committed:
Merge branch 'add/webp-uploads' into add/edit-restore-multiple-mime-types

mehulkaklotar committed:
phpdoc description added for variable

mehulkaklotar committed:
wp_upload_image_mime_transforms takes single parameter fix

mehulkaklotar committed:
use wp_get_default_extension_for_mime_type to get extension for mime type

audrasjb committed:
Docs: Refine `@return` docblock mentions for `esc_sql()`, `wp_slash()` and `wp_unslash()`.

Props malthert, SergeyBiryukov, johnbillion, desrosj, byohann6.
Fixes #53946.
See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53775 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Remove an obsolete function description in `addslashes_gpc()`.

- Usage of `get_magic_quotes_gpc()` was removed from `addslashes_gpc()` in [46105] / #47783.
- A similar description fragment was removed from `wp_parse_str()` in that changeset, but this one was missed.
- The `get_magic_quotes_gpc()` function itself was deprecated in PHP 7.4 and removed in PHP 8.0.

Props mitweka, audrasjb, SergeyBiryukov.
Fixes #56233.
See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53776 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Media: Prevent URLs from overflowing their container in the media editor.

This changeset also introduces the `.word-wrap-break-word` class which can be used to apply `word-wrap: break-word` to admin elements when needed.

Props mitogh, costdev, kapilpaul, alamgircsebd, sabernhardt, anantajitjg, afercia, circlecube, SergeyBiryukov, rafiahmedd, audrasjb.
Fixes #55393.


git-svn-id: https://develop.svn.wordpress.org/trunk@53777 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Revisions: Rename the function for retrieving the latest revision ID and total count.

The new name is `wp_get_latest_revision_id_and_total_count()`.

This aims to reduce ambiguity about what exactly is the "first" or "last" revision, and bring more consistency with similar wording elsewhere in core, e.g. latest posts, latest comments, etc.

Follow-up to [53759], [53769].

Props peterwilsoncc.
See #55857.

git-svn-id: https://develop.svn.wordpress.org/trunk@53778 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Revisions: Update the "last revision" wording to "latest revision" in various files.

This aims to reduce ambiguity about what exactly is the "first" or "last" revision, and bring more consistency with similar wording elsewhere in core, e.g. latest posts, latest comments, etc.

This affects:
* `wp_save_post_revision()`
* `wp_prepare_revisions_for_js()`
* `WP_Customize_Manager::filter_revision_post_has_changed()`

Follow-up to [53759], [53769], [53778].

Props peterwilsoncc.
Fixes #55857.

git-svn-id: https://develop.svn.wordpress.org/trunk@53779 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Docs: Improve parameter descriptions in `render_block` and `render_block_*` filters.

The former `$block_content` parameter descriptions were too opinionated as block content can be appended, prepended, nested, removed, etc.

Props milana_cap.
Fixes #56286.
See #55646.


git-svn-id: https://develop.svn.wordpress.org/trunk@53780 602fd350-edb4-49c9-b593-d223f7449a82

audrasjb committed:
Taxonomy: Allow filtering Ajax term search results in quick edit.

This changeset introduces the `ajax_term_search_results` hook which can be used to filter the term search results returned by the AJAX term query.

Props grandeljay, costdev, ironprogrammer, audrasjb, SergeyBiryukov.
Fixes #55606.


git-svn-id: https://develop.svn.wordpress.org/trunk@53781 602fd350-edb4-49c9-b593-d223f7449a82

mehulkaklotar committed:
use editor/img class for saving full sized images fix

SergeyBiryukov committed:
Tests: Bring some modernization to `wp_insert_post()` tests.

Includes:
* Using `assertIsInt()` instead of `assertIsNumeric()` for post IDs.
* Using consistent variable names for post data and retrieved posts.
* Removing added filters before performing assertions, not after.
* Removing unused `Tests_Post::$post_ids` property.
* Correcting the order of arguments in some assertions.
* Converting some `foreach()` loops to data providers.
* Wrapping long lines for better readability.

Follow-up to [13/tests], [14/tests], [167/tests], [496/tests], [1174/tests], [1246/tests], [1287/tests], [1307/tests], [1326/tests], [30510], [33261], [33630], [34762], [46279], [50012], [51438].

See #55652.

git-svn-id: https://develop.svn.wordpress.org/trunk@53782 602fd350-edb4-49c9-b593-d223f7449a82

mehulkaklotar committed:
unused variables removed

mehulkaklotar committed:
tests fixed according to feedback

SergeyBiryukov committed:
Tests: Move `wp_publish_post()` tests to their own file.

Now that there is a separate test class for `wp_publish_post()` tests, some of the pre-existing tests from the general `Tests_Post` class can be moved there.

Follow-up to [1039/tests], [1174/tests], [46969], [49000].

See #55652.

git-svn-id: https://develop.svn.wordpress.org/trunk@53783 602fd350-edb4-49c9-b593-d223f7449a82

mehulkaklotar committed:
fix code flow and unused conditions refactor

mehulkaklotar committed:
IMAGE_EDIT_OVERWRITE for deleting source files

mehulkaklotar committed:
fix thumbnail edit issue

mehulkaklotar committed:
delete backup sources when attachment is deleted

mehulkaklotar committed:
delete backup sources when attachment is deleted tests added

mehulkaklotar commented:
**Tests**:

- Rotate image and insert image into post.
  - All the images gets rotated when viewing the post as well as in file system. If WebP image is not available for the particular size, it displays the jpeg version.
- Edit same image with another rotation
  - Same results as first test.
- Restore original image 
  - it restores the original images from backup sources and updates the meta as well.
- Deleting an image which is edited.
  - It deleted the images of all mimes. For this we needed to get changes from this [PR #411](https://github.com/WordPress/performance/pull/411/files)
- Post meta and files in system
  - All metadata and files in the system is being removed and cleared.

One major thing, I am noticing is when we upload the image, it generates the WebP version of the files with `-jpg` suffix, but when we edit the image, it doesn't add the `-jpg` suffix but it just uses the original filename instead.
<img width="312" alt="Screenshot 2022-07-26 at 11 39 23 PM" src="https://user-images.githubusercontent.com/4090336/181080528-25f126a1-b238-4722-92cd-94c0dc6cfa1e.png">

adamsilverstein commented:
Thanks for doing all of the testing/validation on the functionality.

> One major thing, I am noticing is when we upload the image, it generates the WebP version of the files with -jpg suffix, but when we edit the image, it doesn't add the -jpg suffix but it just uses the original filename instead.

Good catch.

I will look into this part to see where the difference is. The main reason we decided to add the extension for core is the potential for naming conflicts for output files. For example if a user uploads "cat.jpeg" and later "cat.jpg" - both valid but different JPEG file names - the output webp could overwrite the previous version (eg if both generated the thumbnail "cat-150x150.webp"). By including the original file extension as a suffix in the filename we ensure unique file names for this case.

Based on your description, this is not work correctly when editing, I'll check to see why.

gziolo committed:
Site Editor: Pass correct context into get_block_categories

Site Editor passed incorrectly global $post as a context when setting up block categories. Instead, it should pass $block_editor_context, since global $post is null in this file.

Props Mamaduka.
Fixes #56284.



git-svn-id: https://develop.svn.wordpress.org/trunk@53784 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Tests: Move `wp_insert_post()` tests to their own file.

Now that there is a separate test class for `wp_insert_post()` tests, some of the pre-existing tests from the general `Tests_Post` class can be moved there.

Includes:
* Removal of unnecessarily setting the current user to Editor for all tests.
* Removal of unnecessarily setting the cron option to an empty array for all tests.

Follow-up to [496/tests], [36607], [53782], [53783].

See #55652.

git-svn-id: https://develop.svn.wordpress.org/trunk@53785 602fd350-edb4-49c9-b593-d223f7449a82

felixarntz committed:
Media: Add information about additional MIME type sources to attachments REST endpoints.

This changeset is a follow-up to [53751] which ensures the additional `sources` information stored in attachment metadata are available under `media_details` for each image size in the REST API responses for attachments.

Props mukesh27, eugenemanuilov, mitogh, flixos90, aaemnnosttv.
See #55443.


git-svn-id: https://develop.svn.wordpress.org/trunk@53786 602fd350-edb4-49c9-b593-d223f7449a82

adamsilverstein committed:
Merge branch 'trunk' into add/edit-restore-multiple-mime-types

# Conflicts:
#	src/wp-includes/post.php
#	tests/phpunit/tests/image/editor.php

adamsilverstein committed:
Merge branch 'trunk' into add/edit-restore-multiple-mime-types

adamsilverstein committed:
use img->make_subsize of available

adamsilverstein committed:
use img->generate_filename to generate output file names

adamsilverstein commented:
@mehulkaklotar - I brought this PR up to date with trunk and resolved conflicts cleanly
* I switched the approach to naming the files for save, using `$img->generate_filename`
* I switched the approach to sub-sized generation to use "make_subsize" when available

Please review this screencast that shows my media library and the file system for the current month. I go thru upload, edit and delete. At the end of the process you can see a couple of files are left over, so something is not being stored properly in meta for the cleanup on delete (note I saw the same thing before my changes). 

Can you please review my recent changes and investigate the files not being deleted? I also saw issues with files not being cleaned up (maybe same issue) with large images that got a -scaled extension and after doing an edit and then restore.

adamsilverstein commented:
@mehulkaklotar lets take the conversation to the new PR and close this one, i see what your saying with the filenames, lets work on fixing that in the other PR.  I'll double check the deleting, maybe I have something locally that is doing something unexpected.

adamsilverstein commented:
Missing screencast:


https://user-images.githubusercontent.com/2676022/181341516-85d8125e-c20d-4282-b196-4b22dbc478cc.mp4

adamsilverstein commented:
Closing in favor of https://github.com/WordPress/wordpress-develop/pull/3030
